source/solvers/vof.cc

```diff
@@ -295,6 +295,55 @@ VOF<dim>::calculate_L2_error()
   return l2error;
 }
 
+template <int dim>
+double
+VOF<dim>::calculate_volume(int fluid_index)
+{
+  auto mpi_communicator = triangulation->get_communicator();
+
+  FEValues<dim> fe_values_fs(*this->fs_mapping,
+                             *fe,
+                             *this->error_quadrature,
+                             update_values | update_gradients |
+                               update_quadrature_points | update_JxW_values);
+
+  const unsigned int  dofs_per_cell = fe->dofs_per_cell;
+  const unsigned int  n_q_points    = this->error_quadrature->size();
+  std::vector<double> q_scalar_values(n_q_points);
+
+  double volume = 0;
+
+  for (const auto &cell : dof_handler.active_cell_iterators())
+    {
+      if (cell->is_locally_owned())
+        {
+          fe_values_fs.reinit(cell);
+          fe_values_fs.get_function_values(present_solution, q_scalar_values);
+
+          for (unsigned int q = 0; q < n_q_points; q++)
+            {
+              switch (fluid_index)
+                {
+                  case 0:
+                    {
+                      if (q_scalar_values[q] < 0.5)
+                        volume += fe_values_fs.JxW(q) * q_scalar_values[q];
+                      break;
+                    }
+                  case 1:
+                    {
+                      if (q_scalar_values[q] > 0.5)
+                        volume += fe_values_fs.JxW(q) * q_scalar_values[q];
+                      break;
+                    }
```

Comment:
So you are essentially integrating what is behind in the interface instead of integrating the whole VOF field right? It's a good idea, but don't expect exact conservation here for sure.

Reply:
Yes that's it! I took the usual definition of the literature (the interface is considered located at `phase_value = 0.5`). Right now, with the diffusivity, conservation is not exact (as you say), but still gives a good idea of it's order of magnitude. I think it will work well with Shahab's interface sharpening, we will see if another criterion is needed at this point :)